pyccel/ast/variable.py

```diff
@@ -0,0 +1,830 @@
+                       wish to collect
+        new_name     : str
+                       The new name of the variable
+        """
+        assert(not self._is_homogeneous)
+        self._vars[variable_idx].rename(new_name)
+
+    def __getitem__(self, idx):
+        if self._is_homogeneous:
+            return Variable.__getitem__(self, idx)
+        else:
+            if isinstance(idx, tuple):
+                sub_idx = idx[1:]
+                idx = idx[0]
+            else:
+                sub_idx = []
+
+            if isinstance(idx, LiteralInteger):
+                idx = idx.p
+            var = self.get_var(idx)
+
+            if len(sub_idx) > 0:
+                return var[sub_idx]
+            else:
+                return var
+
+    def __iter__(self):
+        return self._vars.__iter__()
+
+    def __len__(self):
+        return len(self._vars)
+
+    @property
+    def inconsistent_shape(self):
+        """ Indicates whether all objects in the tuple have the
+        same shape
+        """
+        return self._inconsistent_shape
+
+    @property
+    def is_homogeneous(self):
+        """ Indicates if all objects in the tuple have the
+        same datatype
+        # TODO: Indicates if all objects in the tuple have the
+        same properties
+        """
+        return self._is_homogeneous
+
+    @is_homogeneous.setter
+    def is_homogeneous(self, is_homogeneous):
+        self._is_homogeneous = is_homogeneous
+
+    @Variable.allocatable.setter
+    def allocatable(self, allocatable):
+        if not isinstance(allocatable, bool):
+            raise TypeError('allocatable must be a boolean.')
+        self._allocatable = allocatable
+        for var in self._vars:
+            var.allocatable = allocatable
+
+    @Variable.is_pointer.setter
+    def is_pointer(self, is_pointer):
+        if not isinstance(is_pointer, bool):
+            raise TypeError('is_pointer must be a boolean.')
+        self._is_pointer = is_pointer
+        for var in self._vars:
+            var.is_pointer = is_pointer
+
+    @Variable.is_target.setter
+    def is_target(self, is_target):
+        if not isinstance(is_target, bool):
+            raise TypeError('is_target must be a boolean.')
+        self._is_target = is_target
+        for var in self._vars:
+            var.is_target = is_target
+
+class Constant(ValuedVariable, PyccelAstNode):
+
+    """
+
+    Examples
+    --------
+
+    """
+
+
+
+class IndexedElement(PyccelAstNode):
+
+    """
+    Represents a mathematical object with indices.
+
+    Examples
+    --------
+    >>> from sympy import symbols, Idx
+    >>> from pyccel.ast.core import Variable, IndexedElement
+    >>> i, j = symbols('i j', cls=Idx)
+    >>> A = Variable('A', dtype='int')
+    >>> IndexedElement(A, i, j)
+    IndexedElement(A, i, j)
+    >>> IndexedElement(A, i, j) == A[i, j]
+    True
+    """
+
+    def __new__(
+        cls,
+        base,
+        *args,
+        **kw_args
+        ):
+
```

Comment:
there was a check for base argument here it is no more needed?

Reply:
The check was quite large and referred to lots of sympy types that should only exist in the syntactic stage. But IndexedElement is only used in the semantic stage so I don't see the relevance. The last PR also added a check for the semantic stage to test that base is a variable. But this condition is not strictly necessary. E.g:
```
a = f()[0]
```
We may wish to forbid such syntax, but in that case a neat error should be raised